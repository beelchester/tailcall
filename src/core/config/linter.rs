use std::collections::BTreeSet;

use anyhow::Result;
use convert_case::{Case, Casing};
use serde::{Deserialize, Serialize};

use super::Config;
use crate::core::config::Variant;
use crate::core::macros::MergeRight;
use crate::core::valid::{Valid, Validator};

#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq, schemars::JsonSchema, MergeRight)]
pub enum TextCase {
    #[serde(rename = "camelCase")]
    CamelCase,
    #[serde(rename = "pascalCase")]
    PascalCase,
    #[serde(rename = "snakeCase")]
    SnakeCase,
    #[serde(rename = "screamingSnakeCase")]
    ScreamingSnakeCase,
    #[serde(rename = "allCaps")]
    AllCaps,
}

impl From<TextCase> for Case {
    fn from(text_case: TextCase) -> Self {
        match text_case {
            TextCase::CamelCase => Case::Camel,
            TextCase::PascalCase => Case::Pascal,
            TextCase::SnakeCase => Case::Snake,
            TextCase::ScreamingSnakeCase => Case::ScreamingSnake,
            TextCase::AllCaps => Case::Upper,
        }
    }
}

#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq, schemars::JsonSchema, MergeRight)]
pub struct Linter {
    pub default: Option<bool>,
    #[serde(rename = "autoFix")]
    pub autofix: Option<bool>,
    #[serde(rename = "field")]
    pub field_lint: Option<TextCase>,
    #[serde(rename = "type")]
    pub type_lint: Option<TextCase>,
    #[serde(rename = "enum")]
    pub enum_lint: Option<TextCase>,
    #[serde(rename = "enumValue")]
    pub enum_value_lint: Option<TextCase>,
}

pub fn lint(mut config: Config) -> Result<Config> {
    let lint = config.clone().server.lint;
    if let Some(lint) = lint {
        let autofix = lint.autofix.unwrap_or(false);
        if lint.field_lint.is_some() || lint.default.unwrap_or(false) {
            config = field_lint(config, autofix).to_result()?;
        }
        if lint.type_lint.is_some() || lint.default.unwrap_or(false) {
            config = type_lint(config, autofix).to_result()?;
        }
        if lint.enum_lint.is_some() || lint.default.unwrap_or(false) {
            config = enum_lint(config, autofix).to_result()?;
        }
        if lint.enum_value_lint.is_some() || lint.default.unwrap_or(false) {
            config = enum_value_lint(config, autofix).to_result()?;
        }
    }
    Ok(config)
}

pub fn field_lint(mut config: Config, autofix: bool) -> Valid<Config, String> {
    if let Some(lint) = config.clone().server.lint {
        let config_case = lint.field_lint.unwrap_or(TextCase::CamelCase);
        let target_case = Case::from(config_case);
        let mut errors = Vec::new();
        config.types.iter_mut().for_each(|(_, type_)| {
            for field in type_.fields.clone() {
                let case = field.0.to_case(target_case);
                if case != *field.0 {
                    if autofix {
                        if let Some(field_info) = type_.fields.remove(&field.0) {
                            tracing::warn!(
                                "field {} is renamed to {}",
                                field.0,
                                field.0.to_case(target_case)
                            );
                            type_
                                .fields
                                .insert(field.0.to_case(target_case), field_info);
                        }
                    } else {
                        errors.push(format!(
                            "lint failed for field {}, expected {}",
                            field.0, case
                        ));
                    }
                }
            }
        });
        if !errors.is_empty() {
            return Valid::fail(errors.join("\n"));
        }
    }
    Valid::succeed(config)
}

pub fn type_lint(mut config: Config, autofix: bool) -> Valid<Config, String> {
    if let Some(lint) = config.clone().server.lint {
        let config_case = lint.type_lint.unwrap_or(TextCase::PascalCase);
        let target_case = Case::from(config_case);
        let mut errors = Vec::new();
        config.clone().types.iter_mut().for_each(|(name, _)| {
            if name.to_case(target_case) != *name {
                if autofix {
                    if let Some(type_info) = config.types.remove(name.as_str()) {
                        tracing::warn!("type {} is renamed to {}", name, name.to_case(target_case));
                        config.types.insert(name.to_case(target_case), type_info);
                    }
                } else {
                    errors.push(format!(
                        "lint failed for type {}, expected {}",
                        name,
                        name.to_case(target_case)
                    ));
                }
            }
        });
        if !errors.is_empty() {
            return Valid::fail(errors.join("\n"));
        }
    }
    Valid::succeed(config)
}

pub fn enum_lint(mut config: Config, autofix: bool) -> Valid<Config, String> {
    if let Some(lint) = config.clone().server.lint {
        let config_case = lint.enum_lint.unwrap_or(TextCase::PascalCase);
        let target_case = Case::from(config_case);
        let mut errors = Vec::new();
        config.clone().enums.iter_mut().for_each(|(name, _)| {
            if name.to_case(target_case) != *name {
                if autofix {
                    if let Some(enum_info) = config.enums.remove(name.as_str()) {
                        tracing::warn!("enum {} is renamed to {}", name, name.to_case(target_case));
                        config.enums.insert(name.to_case(target_case), enum_info);
                    }
                } else {
                    errors.push(format!(
                        "lint failed for enum {}, expected {}",
                        name,
                        name.to_case(target_case)
                    ));
                }
            }
        });
    }
    Valid::succeed(config)
}

pub fn enum_value_lint(mut config: Config, autofix: bool) -> Valid<Config, String> {
    if let Some(lint) = config.clone().server.lint {
        let config_case = lint.enum_value_lint.unwrap_or(TextCase::AllCaps);
        let target_case = Case::from(config_case);
        let mut errors = Vec::new();
        config
            .clone()
            .enums
            .iter_mut()
            .for_each(|(enum_name, enum_)| {
                let updated_variants: BTreeSet<_> = enum_
                    .variants
                    .iter()
                    .map(|variant| {
                        let case_name = variant.name.to_case(target_case);
                        if case_name != variant.name {
                            if autofix {
                                tracing::warn!(
                                    "variant {} is renamed to {}",
                                    variant.name,
                                    case_name
                                );
                                // return the variant with the updated name
                                Variant { name: case_name, ..variant.clone() }
                            } else {
                                errors.push(format!(
                                    "lint failed for enum variant {}, expected {}",
                                    variant.name, case_name
                                ));
                                variant.clone()
                            }
                        } else {
                            variant.clone()
                        }
                    })
                    .collect();

                config.enums.remove(enum_name);
                enum_.variants = updated_variants;
                config.enums.insert(enum_name.into(), enum_.clone());
            });
        if !errors.is_empty() {
            return Valid::fail(errors.join("\n"));
        }
    }
    Valid::succeed(config)
}
