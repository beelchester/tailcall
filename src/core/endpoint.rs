use derive_setters::Setters;
use hyper::HeaderMap;

use crate::core::config::Encoding;
use crate::core::http::Method;
use crate::core::json::JsonSchema;

use super::config::Proxy;

#[derive(Clone, Debug, Setters)]
pub struct Endpoint {
    pub path: String,
    pub query: Vec<(String, String, bool)>,
    pub method: Method,
    pub proxy: Option<Proxy>,
    pub input: JsonSchema,
    pub output: JsonSchema,
    pub headers: HeaderMap,
    pub body: Option<String>,
    pub description: Option<String>,
    pub encoding: Encoding,
}

impl Endpoint {
    pub fn new(url: String) -> Endpoint {
        Self {
            path: url,
            query: Default::default(),
            method: Default::default(),
            proxy: Default::default(),
            input: Default::default(),
            output: Default::default(),
            headers: Default::default(),
            body: Default::default(),
            description: Default::default(),
            encoding: Default::default(),
        }
    }
}
