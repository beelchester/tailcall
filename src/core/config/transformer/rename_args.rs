use indexmap::IndexMap;

use crate::core::config::Config;
use crate::core::valid::{Valid, Validator};
use crate::core::Transform;

/// A transformer that renames existing args by replacing them with suggested
/// names.
pub struct RenameArgs(IndexMap<String, Vec<(String, String)>>);

impl RenameArgs {
    pub fn new<I: Iterator<Item = (S, Vec<(S, S)>)>, S: ToString>(suggested_names: I) -> Self {
        Self(
            suggested_names
                .map(|(a, b)| {
                    (
                        a.to_string(),
                        b.into_iter()
                            .map(|(c, d)| (c.to_string(), d.to_string()))
                            .collect(),
                    )
                })
                .collect(),
        )
    }
}

impl Transform for RenameArgs {
    type Value = Config;
    type Error = String;

    fn transform(&self, config: Self::Value) -> Valid<Self::Value, Self::Error> {
        let mut config = config;

        Valid::from_iter(
            self.0.iter(),
            |(field_name, existing_and_suggested_names)| {
                let query_type = config.types.get_mut("Query");
                if let Some(type_) = query_type {
                    if let Some(mut field_info) = type_.fields.remove(field_name) {
                        for (existing_name, suggested_name) in existing_and_suggested_names {
                            let arg_value = field_info.args.remove(existing_name);
                            if let Some(arg_value) = arg_value {
                                field_info
                                    .args
                                    .insert(suggested_name.to_string(), arg_value);
                                type_
                                    .fields
                                    .insert(field_name.to_string(), field_info.clone());
                            } else {
                                return Valid::fail(format!(
                                    "Failed to find argument '{}' in field {}.",
                                    existing_name, field_name
                                ));
                            }
                        }
                        Valid::succeed(())
                    } else {
                        Valid::fail(format!(
                            "Failed to find field '{}' in configuration.",
                            field_name
                        ))
                    }
                } else {
                    Valid::fail("Failed to find Query type in configuration.".to_string())
                }
            },
        )
        .map(|_| config)
    }
}

// #[cfg(test)]
// mod test {
//     use indexmap::IndexMap;
//     use maplit::hashmap;
//
//     use super::RenameArgs;
//     use crate::core::config::Config;
//     use crate::core::transform::Transform;
//     use crate::core::valid::{ValidationError, Validator};
//
//     #[test]
//     fn test_rename_type() {
//         let sdl = r#"
//             schema {
//                 query: Query
//             }
//             type A {
//                 id: ID!
//                 name: String
//             }
//             type Post {
//                 id: ID!
//                 title: String
//                 body: String
//             }
//             type B {
//                 name: String
//                 username: String
//             }
//             type Query {
//                 posts: [Post] @http(path: "/posts")
//             }
//             type Mutation {
//               createUser(user: B!): A @http(method: POST, path: "/users", body: "{{args.user}}")
//             }
//         "#;
//         let config = Config::from_sdl(sdl).to_result().unwrap();
//
//         let cfg = RenameArgs::new(
//             hashmap! {
//                 "Query" => "PostQuery",
//                 "A" => "User",
//                 "B" => "InputUser",
//                 "Mutation" => "UserMutation",
//             }
//             .iter(),
//         )
//         .transform(config)
//         .to_result()
//         .unwrap();
//
//         insta::assert_snapshot!(cfg.to_sdl())
//     }
//
//     #[test]
//     fn test_should_raise_error_when_operation_type_name_is_different() {
//         let sdl = r#"
//             schema {
//                 query: PostQuery
//             }
//             type Post {
//                 id: ID
//                 title: String
//             }
//             type PostQuery {
//                 posts: [Post] @http(path: "/posts")
//             }
//         "#;
//         let config = Config::from_sdl(sdl).to_result().unwrap();
//
//         let result = RenameArgs::new(hashmap! {"Query" =>  "PostQuery"}.iter())
//             .transform(config)
//             .to_result();
//         assert!(result.is_err());
//     }
//
//     #[test]
//     fn test_should_raise_error_when_type_not_found() {
//         let sdl = r#"
//             schema {
//                 query: Query
//             }
//             type A {
//                 id: ID!
//                 name: String
//             }
//             type Post {
//                 id: ID!
//                 title: String
//                 body: String
//             }
//             type Query {
//                 posts: [Post] @http(path: "/posts")
//             }
//         "#;
//         let config = Config::from_sdl(sdl).to_result().unwrap();
//
//         let mut suggested_names = IndexMap::new();
//         suggested_names.insert("Query", "PostQuery");
//         suggested_names.insert("A", "User");
//         suggested_names.insert("B", "User");
//         suggested_names.insert("C", "User");
//
//         let actual = RenameArgs::new(suggested_names.iter())
//             .transform(config)
//             .to_result();
//
//         let b_err = ValidationError::new("Type 'B' not found in configuration.".to_string());
//         let c_err = ValidationError::new("Type 'C' not found in configuration.".to_string());
//         let expected = Err(b_err.combine(c_err));
//         assert_eq!(actual, expected);
//     }
// }
