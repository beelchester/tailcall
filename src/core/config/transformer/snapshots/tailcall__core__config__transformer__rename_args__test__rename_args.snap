---
source: src/core/config/transformer/rename_args.rs
expression: cfg.to_sdl()
---
schema @server @upstream {
  query: Query
  mutation: Mutation
}

type Mutation {
  createUser(userId: Int!): A @http(body: "{{args.userId}}", method: "POST", path: "/users")
}

type Post {
  body: String
  id: ID!
  title: String
}

type Query {
  post(postId: ID!): Post @http(path: "/posts/{{args.postId}}")
}

type User {
  id: ID!
  name: String
}
